post-processing/post_processing.py

```diff
@@ -337,61 +339,45 @@ def read_perflog(path):
-                    # look for field names that match required columns
-                    required_field_matches = [len(list(filter(re.compile(rexpr).match, COLUMN_NAMES))) > 0 for rexpr in REQUIRED_LOG_FIELDS]
-
-                    # check all required columns are present
-                    if False in required_field_matches:
-                        raise KeyError("Perflog missing one or more required fields", REQUIRED_LOG_FIELDS)
-
-                # determine dataframe column names
-                elif fileinput.lineno() == 2:
-
-                    # get display name index
-                    display_name_index = COLUMN_NAMES.index("display_name")
-                    # break up display name into test name and parameters
-                    display_name = columns[display_name_index]
-                    _, params = get_display_name_info(display_name)
-
-                    # remove display name
-                    COLUMN_NAMES.pop(display_name_index)
-                    # replace with test name and params
-                    COLUMN_NAMES[display_name_index:display_name_index] = [name for name in params]
-                    COLUMN_NAMES.insert(display_name_index, "test_name")
-
-                    # store as dictionary
-                    record = dict(zip(COLUMN_NAMES, prepare_columns(columns, display_name_index)))
-                    records.append(record)
-
-                else:
-                    # store as dictionary
-                    record = dict(zip(COLUMN_NAMES, prepare_columns(columns, display_name_index)))
-                    records.append(record)
-
-        except Exception as e:
-            e.args = (e.args[0] + " in file \'{0}\':".format(path),) + e.args[1:]
-            raise
-
-    return pd.DataFrame.from_records(records)
+    # look for required column matches
+    required_field_matches = [len(list(filter(re.compile(rexpr).match, df.columns))) > 0 for rexpr in REQUIRED_LOG_FIELDS]
+    # check all required columns are present
+    if False in required_field_matches:
+        raise KeyError("Perflog missing one or more required fields", REQUIRED_LOG_FIELDS)
+
+    # replace display name
+    results = df["display_name"].apply(get_display_name_info)
+    index = df.columns.get_loc("display_name")
+    # insert new columns and contents
+    insert_key_cols(df, index, [r[1] for r in results])
+    df.insert(index, "test_name", [r[0] for r in results])
+    # drop old column
+    df.drop("display_name", axis=1, inplace=True)
+
+    # replace other columns with dictionary contents
+    dict_cols = [c for c in ["extra_resources", "env_vars"] if c in df.columns]
+    for col in dict_cols:
+        results = df[col].apply(lambda x: ast.literal_eval(x))
+        # insert new columns and contents
+        insert_key_cols(df, df.columns.get_loc(col), results)
+        # drop old column
+        df.drop(col, axis=1, inplace=True)
+
+    # infer numeric columns
+    df = df.apply(pd.to_numeric, errors="ignore")

```

Comment:
Date filtering works as expected, but different types of missing values result in different type inference behaviour from pandas. Columns that contain entirely `null`, `nan`, `NaN`, or `""` (empty field) are typed as float; everything else (`Null`, `NAN`, `none`, `None`) is typed as object (string). If you have a mixture of the two types of missing values (e.g. `null` and `Null`), automatic typing defaults to object.

It gets more interesting when you put regular values into the mix:
- A mixture of numeric and object-typed missing values will resolve as an object typed column.
- A mixture of string and float-typed missing values will resolve as an object typed column.
- A mixture of numeric and float-typed missing values will resolve as a float typed column.

Reply:
This is getting too complicated. It would be more straightforward to add type in the config file for both the axis, all the filter columns, and all the series columns. And enforce that type here. the string->float will most likely need a custom function, in order to avoid type-errors.

Reply:
Float columns allow for missing values, but int columns don't. For now I'll default to filling missing values in to-be-int columns with 0. Later we could perhaps let the user specify a fill value.

Reply:
Yes, use Int64 instead. But don't rely on the user to get the right type, change it internally.